pyccl/pk2d.py

```diff
@@ -194,71 +227,79 @@ def eval(self, k, a, cosmo):
                 covered by the arrays you passed when initializing this object,
                 the power spectrum will be extrapolated from the earliest
                 available value using the linear growth factor (for which a
-                cosmology is needed).
+                cosmology is needed). If no Cosmology is passed, attempting
+                to evaluate the power spectrum outside of the scale factor
+                boundaries will raise an exception.
 
         Returns:
             float or array_like: value(s) of the power spectrum.
         """
-        # make sure we have growth factors for extrapolation
-        cosmo.compute_growth()
+        # determine if logarithmic derivative is needed
+        if not derivative:
+            eval_funcs = lib.pk2d_eval_single, lib.pk2d_eval_multi
+        else:
+            eval_funcs = lib.pk2d_der_eval_single, lib.pk2d_der_eval_multi
 
-        status = 0
-        cospass = cosmo.cosmo
+        # handle scale factor extrapolation
+        if cosmo is None:
+            cosmo = self.eval._cosmo
+            self.psp.extrap_linear_growth = 404  # flag no extrapolation
+        else:
+            cosmo.compute_growth()  # growth factors for extrapolation
+            self.psp.extrap_linear_growth = 401  # flag extrapolation
 
+        status = 0
         if isinstance(k, int):
             k = float(k)
         if isinstance(k, float):
-            f, status = lib.pk2d_eval_single(self.psp, np.log(k), a, cospass,
-                                             status)
+            f, status = eval_funcs[0](self.psp, np.log(k), a,
+                                      cosmo.cosmo, status)
         else:
             k_use = np.atleast_1d(k)
-            f, status = lib.pk2d_eval_multi(self.psp, np.log(k_use),
-                                            a, cospass,
-                                            k_use.size, status)
+            f, status = eval_funcs[1](self.psp, np.log(k_use), a,
+                                      cosmo.cosmo, k_use.size, status)
+
+        # Catch scale factor extrapolation bounds error.
+        if status == lib.CCL_ERROR_SPLINE_EV:
+            raise TypeError(
+                "Pk2D evaluation scale factor is outside of the "
+                "interpolation range. To extrapolate, pass a Cosmology.")
         check(status, cosmo)
-
         return f
 
+    # Save a dummy cosmology as an attribute of the `eval` method so we don't
+    # have to initialize one every time no `cosmo` is passed. This is gentle
+    # with memory too, as `free` does not work for an empty cosmology.
+    eval._cosmo = type("Dummy", (object,), {"cosmo": lib.cosmology()})()
```

Comment:
1. If we pass `None` it kills the kernel whenever `a < amin`, so we definitely have to pass a `ccl_cosmology` object.
`double ccl_f2d_t_eval(ccl_f2d_t *f2d,double lk,double a,void *cosmo, int *status) {...`
2. `lib.cosmology` creates an empty cosmo struct, same type as the one in `ccl.Cosmology.cosmo`.
3. In order to not create dummy cosmologies every time, I create it once, at import, and just use that one if we don't need extrapolation.
4. It doesn't leak memory because this `lib.cosmology` object is only created once, and lives for the duration of that `pyccl` import. It automatically gets purged when we exit the kernel. It *would* leak if we placed it inside of the `__call__` method, but we aren't.

Reply:
Originally I had put it inside of the `__call__` method and was trying to free the resources with `cosmology_free` but this doesn't work if the struct is empty.

Reply:
ok, let's keep it as it is for now then